src/distrs/beta.jl

```diff
@@ -1,19 +1,71 @@
 # functions related to beta distributions
 
+using HypergeometricFunctions: _₂F₁
+
 import .RFunctions:
-    betapdf,
-    betalogpdf,
-    betacdf,
-    betaccdf,
-    betalogcdf,
-    betalogccdf,
-    betainvcdf,
-    betainvccdf,
+    # betapdf,
+    # betalogpdf,
+    # betacdf,
+    # betaccdf,
+    # betalogcdf,
+    # betalogccdf,
+    # betainvcdf,
+    # betainvccdf,
     betainvlogcdf,
     betainvlogccdf
 
 # pdf for numbers with generic types
-betapdf(α::Real, β::Real, x::Number) = x^(α - 1) * (1 - x)^(β - 1) / beta(α, β)
+betapdf(α::Real, β::Real, x::Real) = exp(betalogpdf(α, β, x))
 
 # logpdf for numbers with generic types
-betalogpdf(α::Real, β::Real, x::Number) = (α - 1) * log(x) + (β - 1) * log1p(-x) - logbeta(α, β)
+function betalogpdf(α::T, β::T, x::T) where T<:Real
+    return if x <= 0.5
+        xlogy(α - 1, x) + (β - 1) * log1p(-x) - logbeta(α, β)
+    else
+        betalogpdf(β, α, 1 - x)
+    end
+end
+betalogpdf(α::Real, β::Real, x::Real) = betalogpdf(promote(α, β, x)...)
+
+betacdf(α::Float64, β::Float64, x::Float64) = first(beta_inc(α, β, x))
+betacdf(α::Real, β::Real, x::Real) = betacdf(promote(float(α), β, x)...)
+betacdf(α::T, β::T, x::T) where T = throw(MethodError(betacdf, (α, β, x)))
+
+betaccdf(α::Float64, β::Float64, x::Float64) = last(beta_inc(α, β, x))
+betaccdf(α::Real, β::Real, x::Real) = betaccdf(promote(float(α), β, x)...)
+betaccdf(α::T, β::T, x::T) where T = throw(MethodError(betaccdf, (α, β, x)))
+
+# The log version is currently based on non-log version. When the cdf is very small we shift
+# to a implementation based on the hypergeometric function ₂F₁ to avoid underflow.
+function betalogcdf(α::Float64, β::Float64, x::Float64)
+    p, q = beta_inc(α, β, x)
+    if p < eps(one(p))
+        # see https://dlmf.nist.gov/8.17#E7
+        return -log(α) + α*log(x) + log(_₂F₁(promote(α, 1 - β, α + 1, x)...)) - logbeta(α, β)
+    elseif p <= 0.7
+        return log(p)
+    else
+        return log1p(-q)
+    end
+end
+betalogcdf(α::Real, β::Real, x::Real) = betalogcdf(promote(float(α), β, x)...)
+betalogcdf(k::T, x::T) where T = throw(MethodError(betalogcdf, (α, β, x)))
```

Comment:
The idea was that the current implementations are restricted to `Float64` already and that we don't have generic versions available right now so that aspect should be handled separately anyway. I'd be okay with moving some of the promotion to SpecialFunctions though